From 8d705becfe69643e2e11c3252b309500b8a74607 Mon Sep 17 00:00:00 2001
From: Anoop C S <anoopcs@samba.org>
Date: Tue, 22 Sep 2026 16:26:05 +0530
Subject: [PATCH] s3/g_lock: Create the stack frame before the fast path

The fast path calls talloc_tos() before a stack frame exists when
invoked by callers such as the smbconf Python bindings. Allocate the
frame before this path and route exits through a common label.

BUG: https://bugzilla.samba.org/show_bug.cgi?id=16258

Signed-off-by: Anoop C S <anoopcs@samba.org>
Reviewed-by: Volker Lendecke <vl@samba.org>
(cherry picked from commit fdf6726c1c6f7274675f070c0d0761ea5101fd4b)
---
 source3/lib/g_lock.c | 27 ++++++++++++++++-----------
 1 file changed, 16 insertions(+), 11 deletions(-)

diff --git a/source3/lib/g_lock.c b/source3/lib/g_lock.c
index 08e2b2d7aa75..3174151c2fb1 100644
--- a/source3/lib/g_lock.c
+++ b/source3/lib/g_lock.c
@@ -1288,6 +1288,8 @@ NTSTATUS g_lock_lock(struct g_lock_ctx *ctx, TDB_DATA key,
 		return NT_STATUS_INVALID_PARAMETER_5;
 	}
 
+	frame = talloc_stackframe();
+
 	if ((type == G_LOCK_READ) || (type == G_LOCK_WRITE)) {
 		/*
 		 * This is an abstraction violation: Normally we do
@@ -1310,7 +1312,7 @@ NTSTATUS g_lock_lock(struct g_lock_ctx *ctx, TDB_DATA key,
 		if (!NT_STATUS_IS_OK(status)) {
 			DBG_DEBUG("dbwrap_do_locked() failed: %s\n",
 				  nt_errstr(status));
-			return status;
+			goto done;
 		}
 
 		DBG_DEBUG("status=%s, state.status=%s\n",
@@ -1322,19 +1324,23 @@ NTSTATUS g_lock_lock(struct g_lock_ctx *ctx, TDB_DATA key,
 				const char *name = dbwrap_name(ctx->db);
 				dbwrap_lock_order_lock(name, ctx->lock_order);
 			}
-			return NT_STATUS_OK;
+			status = NT_STATUS_OK;
+			goto done;
 		}
 		if (NT_STATUS_EQUAL(state.status, NT_STATUS_WAS_UNLOCKED)) {
 			/* without dbwrap_lock_order_lock() */
-			return NT_STATUS_OK;
+			status = NT_STATUS_OK;
+			goto done;
 		}
 		if (!NT_STATUS_EQUAL(
 			    state.status, NT_STATUS_LOCK_NOT_GRANTED)) {
-			return state.status;
+			status = state.status;
+			goto done;
 		}
 
 		if (timeval_is_zero(&timeout)) {
-			return NT_STATUS_LOCK_NOT_GRANTED;
+			status = NT_STATUS_LOCK_NOT_GRANTED;
+			goto done;
 		}
 
 		/*
@@ -1344,26 +1350,25 @@ NTSTATUS g_lock_lock(struct g_lock_ctx *ctx, TDB_DATA key,
 		 */
 	}
 
-	frame = talloc_stackframe();
 	status = NT_STATUS_NO_MEMORY;
 
 	ev = samba_tevent_context_init(frame);
 	if (ev == NULL) {
-		goto fail;
+		goto done;
 	}
 	req = g_lock_lock_send(frame, ev, ctx, key, type, cb_fn, cb_private);
 	if (req == NULL) {
-		goto fail;
+		goto done;
 	}
 	end = timeval_current_ofs(timeout.tv_sec, timeout.tv_usec);
 	if (!tevent_req_set_endtime(req, ev, end)) {
-		goto fail;
+		goto done;
 	}
 	if (!tevent_req_poll_ntstatus(req, ev, &status)) {
-		goto fail;
+		goto done;
 	}
 	status = g_lock_lock_recv(req);
- fail:
+done:
 	TALLOC_FREE(frame);
 	return status;
 }
